File: libbambuser-ios-0.9.24-191008/testclient/FileUploader.m
/*
 * libbambuser - Bambuser iOS library
 * Copyright 2014 Bambuser AB
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#import <CoreLocation/CoreLocation.h>
#import "FileUploader.h"
#import <sys/utsname.h>
#import <UIKit/UIKit.h>
#import <AssetsLibrary/AssetsLibrary.h>
#import <Photos/Photos.h>

@implementation FileUploader
@synthesize delegate;

-(id) initWithDelegate: (id) _delegate {
	if (self = [super init]) {
		delegate = _delegate;
		uploadURL = nil;
		inputFilename = nil;
	}
	return self;
}

-(NSString *)urlEncode: (NSString*) string {
	return [string stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
}

/*
Keys passed to getTicketAndUpload:
required:
 - fileType (NSString*) - @"image" or @"video"
 - applicationId (NSString*)
 optional:
 - filename (NSString*) - path to file for upload
 - dictionary (NSDictionary*) - custom dictionary to be associated with the upload
 - info (NSDictionary*) - info dictionary for asset
 - asset (ALAsset*) or (PHAsset*) - asset with metadata and reference to image/video.
 - excludeLocationData (NSNumber *) - optional key to disable location parameter

 NOTE: Either filename or asset/info-pair must be supplied.
 NOTE: If applicationId isn't provided, username and password can be supplied for backwards compatibility.
 */
-(void) getTicketAndUpload: (NSDictionary*) assets {
	NSObject *asset = [assets objectForKey: @"asset"];
	NSDictionary *info = [assets objectForKey: @"info"];
	NSString* uploadType = [assets objectForKey: @"fileType"];
	NSString *applicationId = [assets objectForKey: @"applicationId"];
	NSString *username = [assets objectForKey: @"username"];
	NSString *password = [assets objectForKey: @"password"];
	NSString *author = [assets objectForKey: @"author"];
	NSString *title = [assets objectForKey: @"title"];
	NSDictionary *dictionary = [assets objectForKey: @"dictionary"];
	NSDate *created = nil;
	if ([asset isKindOfClass: [ALAsset class]])
		created = [(ALAsset*)asset valueForProperty: ALAssetPropertyDate];
	else if ([asset isKindOfClass: [PHAsset class]])
		created = ((PHAsset*)asset).creationDate;
	NSString *uploadFilename = [assets objectForKey: @"filename"];
	NSString *filename;
	NSNumber *deleteAfterUpload = [assets objectForKey: @"deleteAfterUpload"];
	NSNumber *excludeLocationData = [assets objectForKey: @"excludeLocationData"];
	// Keep a reference to the asset info, if available. For assets exported
	// from UIImagePickerController, the given path might become inaccessible when
	// this is released.
	infoDictionary = info;

	if (deleteAfterUpload == nil) {
		deleteOnCompletion = NO;
	} else {
		deleteOnCompletion = [deleteAfterUpload boolValue];
	}

	if (!uploadFilename) {
		if ([info valueForKey: UIImagePickerControllerMediaURL]) {
			uploadFilename = [[info valueForKey: UIImagePickerControllerMediaURL] path];
		}
		if (created) {
			NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
			dateFormatter.dateFormat = @"yyyy-MM-dd'T'HH.mm.ss";
			filename = [dateFormatter stringFromDate: created];
		} else {
			filename = uploadType;
		}
		filename = [filename stringByAppendingString: [uploadType isEqualToString: @"image"] ? @".jpg" : @".mp4"];
	} else {
		// This properly handles both plain file paths and file:// urls; previously
		// it only worked if it was a file:// url.
		uploadFilename = [[NSURL URLWithString: uploadFilename] path];
		filename = [uploadFilename lastPathComponent];
	}

	NSMutableDictionary* params = [NSMutableDictionary dictionary];
	[params setValue:uploadType forKey:@"type"];
	[params setValue:filename forKey:@"filename"];

	// optional: custom_data
	if (dictionary) {
		NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dictionary options:NSJSONWritingPrettyPrinted error:nil];
		NSString *custom_data = [[NSString alloc] initWithData: jsonData encoding:NSUTF8StringEncoding];
		[params setValue:custom_data forKey:@"custom_data"];
	}

	// optional: created
	if (created) {
		[params setValue:[NSString stringWithFormat:@"%d", (int) [created timeIntervalSince1970]] forKey:@"created"];
	}

	// optional: author
	if (author) {
		[params setValue:author forKey:@"author"];
	}

	// optional: title
	if (title) {
		[params setValue:title forKey:@"title"];
	}

	// optional: device_model
	struct utsname u;
	uname(&u);
	NSString *device_model = [NSString stringWithFormat:@"%s", u.machine];
	[params setValue:device_model forKey:@"device_model"];

	// optional: client_version
	NSString *bundleIdentifier = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleIdentifier"];
	NSString *bundleVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
	NSString *client_version = [NSString stringWithFormat:@"%@ %@", bundleIdentifier, bundleVersion];
	[params setValue:client_version forKey:@"client_version"];

	// optional: platform
	[params setValue:@"iOS" forKey:@"platform"];

	// optional: platform_version
	NSString *platform_version = [[UIDevice currentDevice] systemVersion];
	[params setValue:platform_version forKey:@"platform_version"];

	// optional: manufacturer
	[params setValue:@"Apple" forKey:@"manufacturer"];

	// optional: latitude, longitude
	CLLocation *coord = nil;
	if ([asset isKindOfClass: [ALAsset class]])
		coord = [(ALAsset*)asset valueForProperty: ALAssetPropertyLocation];
	else if ([asset isKindOfClass: [PHAsset class]])
		coord = ((PHAsset*)asset).location;
	if (coord && ![excludeLocationData boolValue]) {
		[params setValue:[NSString stringWithFormat:@"%f", (float) coord.coordinate.latitude] forKey:@"lat"];
		[params setValue:[NSString stringWithFormat:@"%f", (float) coord.coordinate.longitude] forKey:@"lon"];
	}

	NSData *postData;
	NSString *contentType;
	NSString *authValue = nil;
	NSURL *url;
	NSDictionary *extraHeaders = nil;
	if (applicationId) {
		url = [NSURL URLWithString:@"https://cdn.bambuser.net/uploadTickets"];
		postData = [NSJSONSerialization dataWithJSONObject:params options:NSJSONWritingPrettyPrinted error:nil];
		contentType = @"application/json";
		extraHeaders = @{
			@"X-Bambuser-ApplicationId": applicationId,
			@"Accept": @"application/vnd.bambuser.cdn.v1+json",
			@"X-Bambuser-ClientVersion": client_version,
			@"X-Bambuser-ClientPlatform": [NSString stringWithFormat:@"iOS %@", platform_version]
		};
	} else {
		url = [NSURL URLWithString:@"https://api.bambuser.com/file/ticket.json"];
		NSString *post = @"";
		for (NSString* key in params) {
			NSString* value = [params valueForKey: key];
			if (![post isEqualToString: @""])
				post = [post stringByAppendingString: @"&"];
			post = [post stringByAppendingString: [NSString stringWithFormat:@"%@=%@", key, [self urlEncode: value]]];
		}
		contentType = @"application/x-www-form-urlencoded";

		postData = [post dataUsingEncoding:NSUTF8StringEncoding allowLossyConversion:YES];

		NSString *authStr = [NSString stringWithFormat:@"%@:%@", username, password];
		NSData *authData = [authStr dataUsingEncoding:NSUTF8StringEncoding];
		authValue = [authData base64EncodedStringWithOptions: 0];
	}
	NSString *postLength = [NSString stringWithFormat:@"%lu", (unsigned long)[postData length]];

	NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
	[request setURL: url];
	[request setHTTPMethod:@"POST"];
	[request setTimeoutInterval: 10.0f];
	if (authValue)
		[request setValue: [NSString stringWithFormat: @"Basic %@",authValue] forHTTPHeaderField: @"Authorization"];
	[request setValue: postLength forHTTPHeaderField: @"Content-Length"];
	[request setValue: contentType forHTTPHeaderField: @"Content-Type"];
	if (extraHeaders) {
		for (NSString* key in extraHeaders)
			[request setValue: [extraHeaders valueForKey: key] forHTTPHeaderField: key];
	}
	[request setHTTPBody: postData];

	ticketTask = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
		if (error != nil) {
			if ([self->delegate respondsToSelector:@selector(showError:)]) {
				[self->delegate performSelectorOnMainThread: @selector(showError:) withObject: [error localizedDescription] waitUntilDone:NO];
			}
			NSLog(@"%@", error);
			return;
		}

		NSDictionary *parsedData = [NSJSONSerialization JSONObjectWithData: data options:0 error:&error];
		if (!parsedData) {
			if ([self->delegate respondsToSelector:@selector(showError:)]) {
				[self->delegate performSelectorOnMainThread: @selector(showError:) withObject:[[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding] waitUntilDone:NO];
			}
			return;
		}

		if ([parsedData valueForKey: @"error"]) {
			NSLog(@"%@", [parsedData valueForKey: @"error"]);
			if ([self->delegate respondsToSelector:@selector(showError:)]) {
				[self->delegate performSelectorOnMainThread: @selector(showError:) withObject:[NSString stringWithFormat: @"%@", [[parsedData valueForKey: @"error"] valueForKey: @"message"]] waitUntilDone:NO];
			}
			return;
		}

		self->uploadURL = [NSURL URLWithString: [parsedData valueForKey:@"upload_url"]];
		if (self->uploadURL == nil) {
			if ([self->delegate respondsToSelector:@selector(showError:)]) {
				[self->delegate performSelectorOnMainThread: @selector(showError:) withObject: @"Unexpected response from upload server, please try again." waitUntilDone:NO];
			}
			return;
		}

		NSString *localUploadFilename = uploadFilename;
		if (!localUploadFilename) {
			if ([uploadType isEqual: @"image"] && info != nil) {
				localUploadFilename = [NSTemporaryDirectory() stringByAppendingPathComponent: @"image.jpg"];
				[UIImageJPEGRepresentation([info valueForKey: UIImagePickerControllerOriginalImage], 1.0) writeToFile: localUploadFilename atomically: YES];
			}
		}

		if (localUploadFilename) {
			[self performSelectorOnMainThread: @selector(uploadFile:) withObject: localUploadFilename waitUntilDone: YES];
			return;
		}
		if ([self->delegate respondsToSelector:@selector(uploadFailed)]) {
			[self->delegate performSelectorOnMainThread:@selector(uploadFailed) withObject:nil waitUntilDone:NO];
		}
	}];
	[ticketTask resume];
}

/*
Alternative upload entrypoint, for when you have an upload ticket instead of an applicationId

Example scenario: Your own server holds the Bambuser credentials and hands out upload tickets
on demand to your iOS app.
*/
- (void) uploadFile:(NSString *) filename toUploadURL:(NSURL *)ticketUploadURL {
	uploadURL = [ticketUploadURL copy];
	[self uploadFile:filename];
}

- (void) uploadFile: (NSString*) filename {
	inputFilename = [filename copy];
	NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
	[request setURL: uploadURL];
	[request setHTTPMethod: @"PUT"];

	NSFileManager *man = [NSFileManager defaultManager];
	filesize = [[man attributesOfItemAtPath: filename error: NULL] fileSize];
	responseStatus = 0;

	[request addValue: [NSString stringWithFormat: @"%"PRId64"", filesize] forHTTPHeaderField: @"Content-Length"];

	[request setHTTPBodyStream: [[NSInputStream alloc] initWithFileAtPath: filename]];
	[request setTimeoutInterval: 60.0f];

	NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
	NSOperationQueue *queue = [[NSOperationQueue alloc] init];
	NSURLSession *session = [NSURLSession sessionWithConfiguration:sessionConfig delegate:self delegateQueue:queue];
	uploadTask = [session uploadTaskWithStreamedRequest:request];
	[uploadTask resume];
	if ([delegate respondsToSelector:@selector(uploadStarted:)]) {
		[delegate performSelectorOnMainThread:@selector(uploadStarted:) withObject:[NSNumber numberWithBool: YES] waitUntilDone:NO];
	}
}

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task needNewBodyStream:(void (^)(NSInputStream *bodyStream))completionHandler {
	NSInputStream* inputStream = [[NSInputStream alloc] initWithFileAtPath: inputFilename];
	completionHandler(inputStream);
}

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didSendBodyData:(int64_t)bytesSent totalBytesSent:(int64_t)totalBytesSent totalBytesExpectedToSend:(int64_t)totalBytesExpectedToSend {
	float progress = ((float)totalBytesSent) / ((float)filesize);
	if ([delegate respondsToSelector:@selector(uploadUpdated:)]) {
		[delegate performSelectorOnMainThread:@selector(uploadUpdated:) withObject: [NSNumber numberWithFloat: progress] waitUntilDone:NO];
	}
}

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error {
	if (error != nil) {
		if (error.code != NSURLErrorCancelled) {
			if ([delegate respondsToSelector:@selector(uploadFailed)])
				[delegate performSelectorOnMainThread:@selector(uploadFailed) withObject:nil waitUntilDone:NO];
		}
	} else {
		if (responseStatus == 200) {
			if ([delegate respondsToSelector:@selector(uploadEnded)]) {
				[delegate performSelectorOnMainThread:@selector(uploadEnded) withObject:nil waitUntilDone:NO];
			}
		} else {
			if ([delegate respondsToSelector:@selector(uploadFailed)]) {
				[delegate performSelectorOnMainThread:@selector(uploadFailed) withObject:nil waitUntilDone:NO];
			}
		}
	}
	[self cancelAndFinalizeUploadTask];
}

- (void)URLSession:(NSURLSession *)session didBecomeInvalidWithError:(NSError *)error {
	if (error != nil) {
		if ([delegate respondsToSelector:@selector(uploadFailed)]) {
			[delegate performSelectorOnMainThread:@selector(uploadFailed) withObject:nil waitUntilDone:NO];
		}
	}
	[self cancelAndFinalizeUploadTask];
}

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data {
	NSLog(@"%@", [[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding]);
}

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveResponse:(NSURLResponse *)response completionHandler:(void (^)(NSURLSessionResponseDisposition disposition))completionHandler {
	NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*)response;
	responseStatus = (int)[httpResponse statusCode];
	if (responseStatus == 200) {
	} else if (responseStatus >= 400 && responseStatus < 500) {
		// Client error
		NSLog(@"Upload was not accepted");
	} else if (responseStatus >= 500 && responseStatus < 600) {
		// Server error
		NSLog(@"Upload was not accepted");
	} else {
		NSLog(@"Unexpected status code: %d", responseStatus);
	}
	completionHandler(NSURLSessionResponseAllow);
}

- (void) cancelUpload {
	[self cancelAndFinalizeUploadTask];
}

- (void) cancelAndFinalizeUploadTask {
	[uploadTask cancel];
	uploadTask = nil;
	[ticketTask cancel];
	ticketTask = nil;
	infoDictionary = nil;
	if (deleteOnCompletion) {
		[[NSFileManager defaultManager] removeItemAtPath: inputFilename error: NULL];
	}
}


@end

File: libbambuser-ios-0.9.24-191008/testclient-swift/FileUploader.swift
/*
 * libbambuser - Bambuser iOS library
 * Copyright 2015 Bambuser AB
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

import Foundation
import AssetsLibrary
import CoreLocation
import Photos

@objc protocol FileUploaderDelegate
{
	@objc optional func uploadStarted(_ async : NSNumber)
	@objc optional func uploadUpdated(_ progress : NSNumber)
	@objc optional func uploadEnded()
	@objc optional func showError(_ errorMessage : String)
	@objc optional func uploadFailed()
}

class FileUploader: NSObject, URLSessionDelegate, URLSessionTaskDelegate, URLSessionDataDelegate {
	weak var delegate: AnyObject?
	var uploadURL: URL?
	var inputFilename: String?
	var ticketTask: URLSessionDataTask?
	var uploadTask: URLSessionUploadTask?
	var deleteOnCompletion: Bool
	var filesize: Int64?
	var responseStatus: Int?
	var infoDictionary: Dictionary<String, AnyObject>?
	init(_delegate: AnyObject) {
		delegate = _delegate
		uploadURL = nil
		inputFilename = nil
		deleteOnCompletion = false
	}

	func urlEncode(_ string: String) -> String {
		let escapedString = string.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)
		return escapedString!
	}

	/*
	Keys passed to getTicketAndUpload:
	required:
 - fileType (NSString*) - @"image" or @"video"
 - applicationId (NSString*)
 optional:
 - filename (NSString*) - path to file for upload
 - dictionary (NSDictionary*) - custom dictionary to be associated with the upload
 - info (NSDictionary*) - info dictionary for asset
 - asset (ALAsset*) or (PHAsset*) - asset with metadata and reference to image/video.
 - library (ALAssetsLibrary*)
 - excludeLocationData (Bool) - optional key to disable location parameter

 NOTE: Either filename or asset/info-pair must be supplied.
 NOTE: If applicationId isn't provided, username and password can be supplied for backwards compatibility.
 */
	func getTicketAndUpload(_ assets: Dictionary <String, AnyObject>) {
		let asset = assets["asset"]
		let info = assets["info"] as? Dictionary <String, AnyObject>
		let uploadType = assets["fileType"] as! String
		let applicationId = assets["applicationId"] as? String
		let username = assets["username"] as? String
		let password = assets["password"] as? String
		let author = assets["author"] as? String
		let title = assets["title"] as? String
		let dictionary = assets["dictionary"] as? Dictionary <String, AnyObject>
		var created:Date?
		if (asset is ALAsset) {
			created = asset!.value(forProperty: ALAssetPropertyDate) as? Date
		} else if (asset is PHAsset) {
			created = (asset as! PHAsset).creationDate
		}
		var uploadFilename = assets["filename"] as? String
		var filename: String
		let deleteAfterUpload = assets["deleteAfterUpload"] as? Bool
		if (deleteAfterUpload == nil) {
			deleteOnCompletion = false
		} else {
			deleteOnCompletion = deleteAfterUpload!
		}
		var excludeLocationData = assets["excludeLocationData"] as? Bool
		if (excludeLocationData == nil) {
			excludeLocationData = false
		}
		// Keep a reference to the asset info, if available. For assets exported
		// from UIImagePickerController, the given path might become inaccessible when
		// this is released.
		infoDictionary = info

		if (uploadFilename == nil) {
			if let mediaurl = info?[UIImagePickerControllerMediaURL] as? URL {
				uploadFilename = mediaurl.path
			}
			if (created != nil) {
				let dateFormatter = DateFormatter()
				dateFormatter.dateFormat = "yyyy-MM-dd'T'HH.mm.ss"
				filename = dateFormatter.string(from: created!)
			} else {
				filename = uploadType
			}
			filename = filename + (uploadType == "image" ? ".jpg" : ".mp4")
		} else {
			// This properly handles both plain file paths and file:// urls; previously
			// it only worked if it was a file:// url.
			uploadFilename = URL(string: uploadFilename!)!.path
			filename = (uploadFilename! as NSString).lastPathComponent
		}

		var params:[String:String] = [String:String]()
		params["type"] = uploadType
		params["filename"] = filename

		// optional: custom_data
		if (dictionary != nil) {
			let jsonData = try? JSONSerialization.data(withJSONObject: dictionary!, options: JSONSerialization.WritingOptions.prettyPrinted)
			let custom_data = NSString(data: jsonData!, encoding: String.Encoding.utf8.rawValue)! as String
			params["custom_data"] = custom_data
		}

		// optional: created
		if (created != nil) {
			params["created"] = String(Int(created!.timeIntervalSince1970))
		}

		// optional: author
		if (author != nil) {
			params["author"] = author!
		}

		// optional: title
		if (title != nil) {
			params["title"] = title!
		}

		// optional: device_model
		var name: [Int32] = [CTL_HW, HW_MACHINE]
		var size: Int = 2
		sysctl(&name, 2, nil, &size, nil, 0)
		var hw_machine = [CChar](repeating: 0, count: Int(size))
		sysctl(&name, 2, &hw_machine, &size, nil, 0)
		let hardware: String = String(cString: hw_machine)
		params["device_model"] = hardware

		// optional: client_version
		let bundleIdentifier = Bundle.main.infoDictionary?["CFBundleIdentifier"] as! String
		let bundleVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as! String
		params["client_version"] = bundleIdentifier + " " + bundleVersion

		// optional: platform
		params["platform"] = "iOS"

		// optional: platform_version
		params["platform_version"] = UIDevice.current.systemVersion

		// optional: manufacturer
		params["manufacturer"] = "Apple"

		//optional: latitude, longitude
		if (asset != nil) {
			var coord: CLLocation?
			if (asset is ALAsset) {
				coord = asset!.value(forProperty: ALAssetPropertyLocation) as? CLLocation
			} else if (asset is PHAsset) {
				coord = (asset as! PHAsset).location
			}

			if (coord != nil && !excludeLocationData!) {
				params["lat"] = coord!.coordinate.latitude.description
				params["lon"] = coord!.coordinate.longitude.description
			}
		}

		let postData:Data?
		let url:URL?
		let contentType:String
		var authValue:String?
		var extraHeaders:[String:String]?
		if (!(applicationId ?? "").isEmpty) {
			url = URL(string: "https://cdn.bambuser.net/uploadTickets")
			postData = try? JSONSerialization.data(withJSONObject: params, options: JSONSerialization.WritingOptions.prettyPrinted)
			contentType = "application/json"

			extraHeaders = [
				"X-Bambuser-ApplicationId": applicationId!,
				"Accept": "application/vnd.bambuser.cdn.v1+json",
				"X-Bambuser-ClientVersion": bundleIdentifier + " " + bundleVersion,
				"X-Bambuser-ClientPlatform": "iOS " + UIDevice.current.systemVersion
			]
		} else {
			url = URL(string: "https://api.bambuser.com/file/ticket.json")
			var post:String = ""
			for (key, value) in params {
				if (post != "") {
					post = post + "&"
				}
				post = post + key + "=" + self.urlEncode(value)
			}
			contentType = "application/x-www-form-urlencoded"

			postData = post.data(using: String.Encoding.utf8, allowLossyConversion: true)

			let authStr = username! + ":" + password!
			let authData = authStr.data(using: String.Encoding.utf8, allowLossyConversion: true)
			authValue = authData?.base64EncodedString(options: NSData.Base64EncodingOptions(rawValue: 0))
		}

		let postLength = postData?.count.description

		let request = NSMutableURLRequest()
		request.url = url
		request.httpMethod = "POST"
		request.timeoutInterval = 10.0
		if (authValue != nil) {
			request.setValue("Basic " + authValue!, forHTTPHeaderField: "Authorization")
		}
		request.setValue(postLength, forHTTPHeaderField: "Content-Length")
		request.setValue(contentType, forHTTPHeaderField: "Content-Type")
		if (extraHeaders != nil) {
			for (key, value) in extraHeaders! {
				request.setValue(value, forHTTPHeaderField: key)
			}
		}
		request.httpBody = postData

		ticketTask = Foundation.URLSession.shared.dataTask(with: request as URLRequest, completionHandler: { (data, response, error) in

			if (error != nil) {
				DispatchQueue.main.async(execute: {
					self.delegate?.showError?(error!.localizedDescription)
				})
				return
			}

			do {
				try JSONSerialization.jsonObject(with: data!, options: .mutableContainers)
			} catch {
				let responseString = NSString(data: data!, encoding:String.Encoding.utf8.rawValue)
				DispatchQueue.main.async(execute: {
					self.delegate?.showError?(responseString! as String)
				})
				return
			}

			let parsedData = (try! JSONSerialization.jsonObject(with: data!, options: JSONSerialization.ReadingOptions())) as! NSDictionary
			if (parsedData["error"] != nil) {
				let responseError = parsedData["error"] as! NSDictionary
				print(responseError)
				let message = responseError["message"]! as! String
				DispatchQueue.main.async(execute: {
					self.delegate?.showError?(message)
				})
				return
			}

			self.uploadURL = URL(string: (parsedData["upload_url"] as? String)!)
			if (self.uploadURL == nil) {
				DispatchQueue.main.async(execute: {
					self.delegate?.showError?("Unexpected response from upload server, please try again.")
				})
				return
			}

			if (uploadFilename == nil) {
				if (uploadType=="image" && info != nil) {
					uploadFilename = (NSTemporaryDirectory() as NSString).appendingPathComponent("image.jpg")

					let image = UIImageJPEGRepresentation(info![UIImagePickerControllerOriginalImage] as! UIImage, 1.0)
					try? image?.write(to: URL(fileURLWithPath: uploadFilename!), options: [.atomic])
				}
			}
			if (uploadFilename != nil) {
				DispatchQueue.main.async(execute: {
					self.uploadFile(uploadFilename!)
				})
				return
			}
			DispatchQueue.main.async(execute: {
				self.delegate?.uploadFailed?()
			})

		})
		ticketTask?.resume()

	}

	/*
	Alternative upload entrypoint, for when you have an upload ticket instead of an applicationId

	Example scenario: Your own server holds the Bambuser credentials and hands out upload tickets
	on demand to your iOS app.
	*/
	func uploadFile(_ filename: String, toUploadURL ticketUploadURL: URL) {
		uploadURL = ticketUploadURL
		uploadFile(filename)
	}

	func uploadFile(_ filename: String) {
		inputFilename = filename
		let request = NSMutableURLRequest()
		request.url = uploadURL!
		request.httpMethod = "PUT"
		let man = FileManager.default
		let attr = (try? man.attributesOfItem(atPath: inputFilename!)) as [FileAttributeKey : AnyObject]?
		filesize = attr?[FileAttributeKey.size]?.int64Value
		responseStatus = 0

		request.addValue(filesize!.description, forHTTPHeaderField: "Content-Length")

		let inputStream = InputStream(fileAtPath: inputFilename!)

		request.httpBodyStream = inputStream
		request.timeoutInterval = 60.0

		let sessionConfig:URLSessionConfiguration = URLSessionConfiguration.default
		let queue = OperationQueue()
		let session:Foundation.URLSession = Foundation.URLSession(configuration: sessionConfig, delegate: self, delegateQueue: queue)
		uploadTask = session.uploadTask(withStreamedRequest: request as URLRequest)
		uploadTask?.resume()
		DispatchQueue.main.async(execute: {
			self.delegate?.uploadStarted?(true)
		})
	}

	func urlSession(_ session: URLSession, task: URLSessionTask, needNewBodyStream completionHandler: @escaping (InputStream?) -> Void) {
		let inputStream = InputStream(fileAtPath: inputFilename!)
		completionHandler(inputStream)
	}

	func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
		let progress = Float(totalBytesSent) / Float(filesize!)
		DispatchQueue.main.async(execute: {
			self.delegate?.uploadUpdated?(NSNumber(value: progress))
		})
	}

	func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
		DispatchQueue.main.async(execute: {
			if (error != nil) {
				if (error!._code != NSURLErrorCancelled) {
					self.delegate?.uploadFailed?()
				}
			} else {
				if (self.responseStatus! == 200) {
					self.delegate?.uploadEnded?()
				} else {
					self.delegate?.uploadFailed?()
				}
			}
		})
		self.cancelAndFinalizeUploadTask()
	}

	func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
		if (error != nil) {
			DispatchQueue.main.async(execute: {
				self.delegate?.uploadFailed?()
			})
		}
		self.cancelAndFinalizeUploadTask()
	}

	func urlSession(_ session: Foundation.URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
		print(NSString(data: data, encoding: String.Encoding.utf8.rawValue)!)
	}

	func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void) {

		if let httpResponse = response as? HTTPURLResponse {
			responseStatus = httpResponse.statusCode

			if (responseStatus == 200) {
			} else if (responseStatus! >= 400 && responseStatus! < 500) {
				// Client error
				print("Upload was not accepted");
			} else if (responseStatus! >= 500 && responseStatus! < 600) {
				// Server error
				print("Upload was not accepted");
			} else {
				print("Unexpected status code:", responseStatus!);
			}
		}
		completionHandler(Foundation.URLSession.ResponseDisposition.allow)
	}

	func cancelUpload() {
		self.cancelAndFinalizeUploadTask()
	}

	fileprivate func cancelAndFinalizeUploadTask() {
		ticketTask?.cancel()
		uploadTask?.cancel()

		ticketTask = nil
		uploadTask = nil
		infoDictionary = nil

		if (deleteOnCompletion) {
			do {
				try FileManager.default.removeItem(atPath: inputFilename!)
			} catch _ {
			}
		}
	}
}
